### PR TITLE
Refactor out SpoonThumbnail

### PR DESCRIPTION
**Pull request by jeroendesloovere**
## Type

- Feature
- Deprecation

## Resolves the following issues

Fixes https://github.com/forkcms/forkcms/issues/708

## Pull request description

Refactor out SpoonThumbnail.
We replaced the functionality with Imagine.

### CODE REVIEW

File: src/ForkCMS/Service/Thumbnails.php [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                $item['width'] = ($chunks[0] !== '') ? (int)$chunks[0] : null;
+                $item['height'] = ($chunks[1] !== '') ? (int)$chunks[1] : null;
+            }
+
+            $return[] = $item;
+        }
+
+        return $return;
+    }
+}
```
Comment (carakas) [inline src/ForkCMS/Service/Thumbnails.php]: new line missing at the end of the file

**Comment by StijnVrolijk** [comment]
I don't like the `Service` namespace. In an ideal (utopian) world, nearly everything should be a service. I'd give this a `Utility` namespace.

**Comment by jeroendesloovere** [comment]
@carakas What do you think of Utils?

**Comment by jeroendesloovere** [comment]
I'm using `Service` because that is the default in Symfony.
Check out this example https://symfony.com/doc/current/service_container.html#creating-configuring-services-in-the-container

**Comment by carakas** [comment]
@jeroendesloovere That is not a default that is an example. The only rule is that the source code needs to be inside src, for the rest you are free (http://fabien.potencier.org/symfony4-directory-structure.html)

File: app/config/config.yml [hunk 2]
```diff
@@ -198,6 +198,9 @@ services:
         tags:
             - { name: kernel.event_listener, event: kernel.response, method: onKernelResponse }
 
+    forkcms.thumbnails:
+        class: ForkCMS\Service\Thumbnails
```
Comment (carakas) [inline app/config/config.yml]: you must declare it public

**Comment by carakas** [comment]
Apart from that is the current recommendation to use the FQCN as service name.
You can always add an alias for `forkcms.utility.thumbnails`

But the benifit is that you can just do $this->getContainer()->get(Thumbnails::class)

File: src/ForkCMS/Service/Thumbnails.php (hunk 1 above)
Comment (jeroendesloovere) [reply src/ForkCMS/Service/Thumbnails.php]: Fixed

**Comment by jeroendesloovere** [comment]
I prefer `forkcms.thumbnails`. @StijnVrolijk what do you think?
We do also use `fork.settings`?

File: app/config/config.yml (hunk 2 above)
Comment (jeroendesloovere) [reply app/config/config.yml]: Fixed

**Comment by StijnVrolijk** [comment]
@jeroendesloovere settings arent really utilities, therefor it makes sense to me to have `fork.settings` but also `forkcms.utility.thumbnails`

**Comment by jeroendesloovere** [comment]
Or should I use the fully qualified name and don't create an alias?

**Comment by carakas** [comment]
I would do both, use the FQCN as service name and forkcms.utility.thumbnails as alias in case somebody wants to use it like that?

**Comment by jeroendesloovere** [comment]
@carakas Can you re-review?

File: app/config/config.yml [hunk 3]
```diff
@@ -198,7 +198,7 @@ services:
         tags:
             - { name: kernel.event_listener, event: kernel.response, method: onKernelResponse }
 
-    forkcms.thumbnails:
+    forkcms.utility.thumbnails:
```
Comment (carakas) [inline app/config/config.yml]:
```yml
ForkCMS\Utility\Thumbnails:
    alias: forkcms.utility.thumbnails
    public: true
```

Can you try this instead

File: bin/console [hunk 4]
```diff
@@ -42,7 +42,7 @@ $backend->initialize('Console');
 
 $application = new Application($kernel);
 $application->add(new \Console\Locale\ImportLocaleCommand());
-$application->add(new \Console\Thumbnails\GenerateThumbnailsCommand());
+$application->add(new \Console\Thumbnails\GenerateThumbnailsCommand($kernel->getContainer()->get('forkcms.utility.thumbnails')));
```
Comment (carakas) [inline bin/console]: maybe you can register it in the yml config instead of hardcoded here?

File: src/Common/Core/Model.php [hunk 5]
```diff
@@ -153,49 +151,26 @@ public static function generatePassword(
...
-            $thumbnail = new \SpoonThumbnail($sourceFile, $folder['width'], $folder['height']);
-            $thumbnail->setAllowEnlargement(true);
-
-            // if the width & height are specified we should ignore the aspect ratio
-            if ($folder['width'] !== null && $folder['height'] !== null) {
-                $thumbnail->setForceOriginalAspectRatio(false);
-            }
-            $thumbnail->parseToFile($folder['path'] . '/' . $filename);
-        }
+        self::get('forkcms.utility.thumbnails')->generate($path, $sourceFile);
```
Comment (carakas) [inline src/Common/Core/Model.php]: when using the class name can also do Classname::class if you want

**Comment by jeroendesloovere** [comment]
Why don't we use `ForkCMS\Utils` instead of `ForkCMS\Utility`?
Symfony always uses `Utils` in their examples, it would be unwise to use another naming then...
https://symfony.com/doc/current/best_practices/business-logic.html

**Comment by StijnVrolijk** [comment]
Utils is an abbreviation of Utilities and abbreviations in code cause confusion (which is the reason we renamed `frm` to `form` for example).
Furthermore we never use plurals in our new namespaces (eg `Command`, `Event`, ..) so I think we should use the singular form: Utility.

**Comment by jeroendesloovere** [comment]
Ok, I searched the web a little and read your opinion and I'm convinced.

Second question.
`ForkCMS\Utility\Thumbnails` or ForkCMS\Utility\GenerateThumbnails`?

File: bin/console (hunk 4 above)
Comment (jeroendesloovere) [reply bin/console]: Done

File: src/Common/Core/Model.php (hunk 5 above)
Comment (jeroendesloovere) [reply src/Common/Core/Model.php]: Done

File: app/config/config.yml (hunk 3 above)
Comment (jeroendesloovere) [reply app/config/config.yml]: Done

**Comment by jeroendesloovere** [comment]
@StijnVrolijk and @carakas can you guys approve?

**Comment by carakas** [comment]
I can approve if the code is good :) I'll see if I still have time today to go through it

File: src/Common/Core/Model.php [hunk 6]
```diff
@@ -153,49 +152,26 @@ public static function generatePassword(
...
+        self::get(Thumbnails::class)->generate($path, $sourceFile);
     }
 
     /**
      * Delete thumbnails based on the folders in the path
      *
      * @param string $path The path wherein the thumbnail-folders exist.
      * @param string|null $thumbnail The filename to be deleted.
+     *
+     * @deprecated Please use the service `forkcms.utility.thumbnails` instead.
```
Comment (carakas) [inline src/Common/Core/Model.php]: you removed the alias from the service definition. You need to update it here or add that alias again

File: src/Console/Thumbnails/GenerateThumbnailsCommand.php [hunk 7]
```diff
@@ -28,41 +35,37 @@ protected function configure(): void
             );
     }
 
-    protected function execute(InputInterface $input, OutputInterface $output): void
+    public function __construct(Thumbnails $thumbnails)
     {
-        // Get input values
-        $folderOption = $input->getOption('folder');
+        $this->finder = new Finder();
```
Comment (carakas) [inline src/Console/Thumbnails/GenerateThumbnailsCommand.php]: I wouldn't move this. It is only used in one method. I'm not sure if the constructor is always called or if there is an optimisation on it. But if it would always be called also when not using this command creating a new instance of finder is a waste of resources

File: src/ForkCMS/Utility/Thumbnails.php [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+     *  - x128 as foldername to generate an image where the height will be
+     *      128px, the width will be calculated based on the aspect ratio.
+     *
+     * @param string $inPath The path wherein the thumbnail-folders will be stored.
+     * @param string $forSourceFile The location of the source file.
+     */
+    public function generate(string $inPath, string $forSourceFile): void
+    {
+        // get folder listing
+        $folders = $this->getFolders($inPath);
```
Comment (carakas) [inline src/ForkCMS/Utility/Thumbnails.php]: why don't you loop over the array that the method returns directly?

File: src/ForkCMS/Utility/Thumbnails.php [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+        foreach ($this->finder->directories()->in($inPath)->name($nameFilter)->depth('== 0') as $directory) {
+            $dirname = $directory->getBasename();
+            $chunks = explode('x', $dirname, 2);
+
+            if (!$includeSourceFolder && count($chunks) !== 2) {
+                continue;
+            }
+
+            $width = (($dirname !== 'source') && ($chunks[0] !== '')) ? (int) $chunks[0] : null;
```
Comment (carakas) [inline src/ForkCMS/Utility/Thumbnails.php]: woulnd't it be better to just check here if the chunk is numerical?

File: src/ForkCMS/Utility/Thumbnails.php [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+            $height = (($dirname !== 'source') && ($chunks[1] !== '')) ? (int) $chunks[1] : null;
```
Comment (carakas) [inline src/ForkCMS/Utility/Thumbnails.php]: same here

File: composer.json [hunk 11]
```diff
@@ -86,5 +86,10 @@
             "**/tests/",
             "**/Test/"
         ]
+    },
+    "autoload-dev": {
```
Comment (carakas) [inline composer.json]: why did you add this one?
Is it a new namespace?

File: src/Common/Core/Model.php (hunk 6 above)
Comment (jeroendesloovere) [reply src/Common/Core/Model.php]: Fixed by 8fd859c

File: src/Console/Thumbnails/GenerateThumbnailsCommand.php (hunk 7 above)
Comment (jeroendesloovere) [reply src/Console/Thumbnails/GenerateThumbnailsCommand.php]: Fixed by c40f720

File: src/ForkCMS/Utility/Thumbnails.php (hunk 8 above)
Comment (jeroendesloovere) [reply src/ForkCMS/Utility/Thumbnails.php]: Fixed

File: src/ForkCMS/Utility/Thumbnails.php (hunk 9 above)
Comment (jeroendesloovere) [reply src/ForkCMS/Utility/Thumbnails.php]: Indeed, changed to `is_int(xxx)`.
Fixed by d1281a4

File: src/ForkCMS/Utility/Thumbnails.php (hunk 10 above)
Comment (jeroendesloovere) [reply src/ForkCMS/Utility/Thumbnails.php]: Indeed, changed to is_int(xxx).
Fixed by d1281a4

File: composer.json (hunk 11 above)
Comment (jeroendesloovere) [reply composer.json]: Tests in Symfony 4 are in `/tests/`.
So I already provided this in the composer.json and placed the tests in the right position.
The namespace `ForkCMS\Tests` was not already existing, so that is probably OK then.

**Comment by jeroendesloovere** [comment]
@carakas Fixed everything in your review.

**Comment by jeroendesloovere** [comment]
@carakas can you re-review this if you find some time :-P

**Comment by carakas** [comment]
If time could be found
Picked up from the ground
The richest on earth
Eternally crowned
